src/org/traccar/database/CommandsManager.java

```diff
@@ -133,4 +127,22 @@ public void sendCommand(Command command, long deviceId, boolean fallbackToText)
         return result;
     }
 
+    private Queue<Command> getDeviceQueue(long deviceId) {
+        if (!deviceQueues.containsKey(deviceId)) {
+            deviceQueues.put(deviceId, new ConcurrentLinkedQueue<Command>());
+        }
+        return deviceQueues.get(deviceId);
+    }
+
+    public void sendQueuedCommands(ActiveDevice activeDevice) {
+        Queue<Command> deviceQueue = deviceQueues.get(activeDevice.getDeviceId());
+        if (deviceQueue != null) {
+            Command command = deviceQueue.poll();
+            while (command != null) {
+                activeDevice.sendCommand(command);
```

Comment:
If exception happen here, command will be lost and rest of queue will be sent when device goes online again.